### PR TITLE
Ensure that shared labels are consistently updated.

### LINKED ISSUE #176: Inconistent formatting when setting labelloc for shared axes
> Ah I see... that is a different bug. 
> 
> To explain what is happening. The axes are shared which makes the left most and bottom most subplot the references. If you set the ticks to be top and right, it updates them for the references. What you want, I think, is set the top labels on for the top and right labels for the right most axes. 
> 
>  

 _Originally posted by @cvanelteren in [#155](https://github.com/Ultraplot/ultraplot/issues/155#issuecomment-2829791466)_

```python
import ultraplot as pplt
fig, axs = pplt.subplots(ncols=2, nrows=2,share='all',refwidth='2cm')
axs.format(xlim=(0,60),ylim=(-20,20), xticklabelloc='t', yticklabelloc='r')
pplt.show()
```

### PR DESCRIPTION
**Pull request by cvanelteren**
Fixes #176 

<details> </summary> code </summary> 

```python
   import ultraplot as uplt, numpy as np

layout = [
    [1, 2, 0],
    [1, 2, 5],
    [3, 4, 5],
    [3, 4, 0],
]
fig, ax = uplt.subplots(layout)

ax.format(
    xticklabelloc="t",
    yticklabelloc="r",
)

layout = [
    [1, 0, 2],
    [0, 3, 0],
    [4, 0, 5],
]

fig, ax = uplt.subplots(layout)
ax.format(
    xticklabelloc="t",
    yticklabelloc="r",
)
uplt.show(block=True)

```
</details>

will now be correct:

image

### CODE REVIEW

**Comment by cvanelteren** [comment]
What I don't like about this fix is that it will fool mpl by turning the `sharex/y` to `None`.  I need to add a way to check for this to turn this back on automatically to prevent a potential edge case to be hit (not sure which).

**Comment by cvanelteren** [comment]
Need to handle non trivial layout

**Review by beckermr — CHANGES_REQUESTED** [review]
The tests here should be used to generate an image comparison plot.

File: ultraplot/tests/test_axes.py [hunk 1]
```diff
@@ -187,3 +187,67 @@ def test_subset_format():
...
+
+    def check_state(numbers: list, state: bool, which: str):
+        for number in numbers:
+            for label in getattr(ax[number], f"get_{which}ticklabels")():
+                assert label.get_visible() == state
+
+    check_state([0, 2], False, which="y")
+    check_state([1, 3, 4], True, which="y")
+    check_state([2, 3], False, which="x")
+    check_state([0, 1, 4], True, which="x")
```
Comment (beckermr) [inline ultraplot/tests/test_axes.py:235]: You didn't test 5.

File: ultraplot/tests/test_axes.py [hunk 2]
```diff
@@ -187,3 +187,67 @@ def test_subset_format():
...
+
+    fig, ax = uplt.subplots(layout)
+    ax.format(
+        xticklabelloc="t",
+        yticklabelloc="r",
+    )
+    check_state([0, 3], False, which="y")
+    check_state([1, 2, 4], True, which="y")
+    check_state([0, 1, 2], True, which="x")
+    check_state([3, 4], False, which="x")
```
Comment (beckermr) [inline ultraplot/tests/test_axes.py:252]: You didn't test 5.

File: ultraplot/tests/test_axes.py [hunk 3]
```diff
@@ -187,3 +187,67 @@ def test_subset_format():
...
+        [4, 0, 5],
+    ]
+
+    fig, ax = uplt.subplots(layout)
+    ax.format(
+        xticklabelloc="t",
+        yticklabelloc="r",
+    )
+    check_state([0, 3], False, which="y")
+    check_state([1, 2, 4], True, which="y")
```
Comment (beckermr) [inline ultraplot/tests/test_axes.py:250]: I don't understand this test case at all.

Comment (beckermr) [reply ultraplot/tests/test_axes.py:250]: I would assume we'd have 2, 0, 5 as on 1, 3, 4 as off for y. What does the plot look like?

Comment (beckermr) [reply ultraplot/tests/test_axes.py:250]: NVM I figured it out sorry.

Comment (beckermr) [inline ultraplot/tests/test_axes.py:250]:
```suggestion
    check_state([0, 1, 4], False, which="y")
    check_state([2, 3, 5], True, which="y")
```

Looking at the example, I would think this test should read like above.

File: ultraplot/tests/test_axes.py [hunk 4]
```diff
@@ -187,3 +187,67 @@ def test_subset_format():
...
+        [0, 3, 0],
+        [4, 0, 5],
+    ]
+
+    fig, ax = uplt.subplots(layout)
+    ax.format(
+        xticklabelloc="t",
+        yticklabelloc="r",
+    )
+    check_state([0, 3], False, which="y")
```
Comment (beckermr) [inline ultraplot/tests/test_axes.py:249]:
```suggestion
    # these correspond to the indices of the axis
    # in the axes array (so the grid number minus 1)
    check_state([0, 3], False, which="y")
```

File: ultraplot/tests/test_axes.py [hunk 5]
```diff
@@ -187,3 +187,67 @@ def test_subset_format():
...
+        xticklabelloc="t",
+        yticklabelloc="r",
+    )
+
+    def check_state(numbers: list, state: bool, which: str):
+        for number in numbers:
+            for label in getattr(ax[number], f"get_{which}ticklabels")():
+                assert label.get_visible() == state
+
+    check_state([0, 2], False, which="y")
```
Comment (beckermr) [inline ultraplot/tests/test_axes.py:232]:
```suggestion
    # these correspond to the indices of the axis
    # in the axes array (so the grid number minus 1)
    check_state([0, 2], False, which="y")
```

**Comment by beckermr** [comment]
An image comparison test is failing, but the artifact has no images in it to look at. I think this is a bug.

**Comment by cvanelteren** [comment]
I am actually tending to use less of the pytest-mpl if possible. These tests take much longer too complete than regular pytests. Currently, the test suite runs over 12 minutes I believe, which is rather long. I think we should move to more selectivel visual tests that benchmark the visual output to reduce the test running times.

**Comment by cvanelteren** [comment]
Need to fix  the edgecase where it is not doing this for alt/twin axes.

**Comment by cvanelteren** [comment]
Note that for twinned axes if set manually this would be the output:

image

This has `twinx` and `twiny` set for `ax[1]` (top right). Not sure if bug or feature.

**Review by beckermr — APPROVED** [review]
